File: src/decoders/timetable-class.ts
import type { SessionHandle, TimetableClass } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { translateToWeekNumber } from "~/api";
import { translatePositionToTimings } from "~/api/helpers/timings";

export const decodeTimetableClass = (item: any, session: SessionHandle): TimetableClass => {
  const startDate = decodePronoteDate(item.DateDuCours.V);
  const blockPosition = item.place;
  const blockLength = item.duree;
  let endDate: Date;

  if (typeof item.DateDuCoursFin?.V === "string") {
    endDate = decodePronoteDate(item.DateDuCoursFin.V);
  }
  else {
    const position = blockPosition % session.instance.blocksPerDay + blockLength - 1;
    const timings = translatePositionToTimings(session, position);

    endDate = new Date(startDate);
    endDate.setHours(timings.hours, timings.minutes);
  }

  return {
    id: item.N,
    backgroundColor: item.CouleurFond,
    notes: item.memo,
    startDate,
    endDate,
    blockLength,
    blockPosition,
    weekNumber: translateToWeekNumber(startDate, session.instance.firstMonday)
  };
};

File: library/src/commonMain/kotlin/decoders/timetable-class.kt
package decoders

import api.helpers.translatePositionToTimings
import api.helpers.translateToWeekNumber
import kotlinx.datetime.*
import kotlinx.serialization.json.*
import models.SessionHandle
import models.TimetableClass

fun decodeTimetableClass (item: JsonObject, session: SessionHandle, decoder: (item: JsonObject) -> Any): TimetableClass<Any> {
    val startDate = decodePronoteDate(item["DateDuCours"]!!.jsonObject["V"]!!.jsonPrimitive.content)
    val blockPosition = item["place"]!!.jsonPrimitive.int
    val blockLength = item["duree"]!!.jsonPrimitive.double
    val endDate: LocalDateTime

    if (item.containsKey("DateDuCoursFin")) {
        if (item["DateDuCoursFin"]!!.jsonObject["V"]!!.jsonPrimitive.isString)
            endDate = decodePronoteDate(item["DateDuCoursFin"]!!.jsonObject["V"]!!.jsonPrimitive.content)
        else {
            val position = blockPosition % session.instance.blocksPerDay + blockLength - 1
            val timings = translatePositionToTimings(session, position.toInt())

            val time = LocalTime(timings.hours.toInt(), timings.minutes.toInt())
            endDate = LocalDateTime(startDate.date, time)
        }
    } else {
        val position = blockPosition % session.instance.blocksPerDay + blockLength - 1
        val timings = translatePositionToTimings(session, position.toInt())

        val time = LocalTime(timings.hours.toInt(), timings.minutes.toInt())
        endDate = LocalDateTime(startDate.date, time)
    }

    return TimetableClass(
        id = item["N"]!!.jsonPrimitive.content,
        backgroundColor = item["CouleurFond"]?.jsonPrimitive?.content,
        notes = item["memo"]?.jsonPrimitive?.content,
        startDate = startDate,
        endDate = endDate,
        blockLength = blockLength,
        blockPosition = blockPosition,
        weekNumber = translateToWeekNumber(startDate, session.instance.firstMonday),
        data = decoder(item)
    )
}